packaging/technical-addon/Splunk_TA_otel/windows_x86_64/bin/Splunk_TA_otel_utils.ps1

```diff
@@ -1,63 +1,88 @@
+		$process = Get-WmiObject Win32_Process | Where-Object ParentProcessId -eq $parentId | Where-Object ProcessName -eq $processName
+		if ($process) {
+			break
+		}
+
+		Start-Sleep -Milliseconds ($sleepInterval.TotalMilliseconds)
+	} while (((Get-Date) - $startTime) -lt $timeout)
+
+	return $process
 }
 
-function waitForExit($parent) 
+function Stop-ProcessById
 {
-	Wait-Process -Id $parent
-	otelLogWrite "INFO Parent process exited"
+	Param ([Int32]$processId)
+
+	Write-Log "INFO Stopping PID $processId if still running"
+	# Use a pipeline to avoid an error message if the process already finished
+	Get-WmiObject Win32_Process | Where-Object ProcessId -eq $processId | ForEach-Object { taskkill.exe /f /pid $_.ProcessId }
 }
 
-function forceStopOtelProcess($processId)
-{
-	Stop-Process -Id $processId -Force
-	Start-Sleep -Seconds 1
+function Get-Win32_Process {
+	Param([Int32]$processId)
+	return Get-WmiObject Win32_Process -Filter "ProcessId = $processId"
 }
 
-function CheckOtelProcessStop($processId)
-{
-	$processInfo = Get-WmiObject win32_process | ? processid -eq $processId
-	if($processInfo) {
-		otelLogWrite "ERROR Otel agent didn't stop"
-	} else {
-		otelLogWrite "INFO Otel agent stopped"
+Write-Log "INFO started"
+try {
+	$currentProcess     = Get-Win32_Process $PID # $PID is predefined as the PID of the current process
+	$parentProcess      = Get-Win32_Process $currentProcess.ParentProcessId
+	$grandParentProcess = Get-Win32_Process $parentProcess.ParentProcessId
+
+	# splunkd launched a cmd that launched the collector, but didn't wait on it and then launched this script
+	# get the expected process for the OTel collector launched via splunkd
+	$otelProcess = Get-Win32_ProcessByNameAndParentId `
+		$otelProcessName `
+		$parentProcess.ProcessId `
+		-timeout ([TimeSpan]::FromSeconds(5)) `
+		-sleepInterval ([TimeSpan]::FromMilliseconds(150))
+	if ($null -eq $otelProcess) {
+		Write-Log "ERROR Otel agent not running"
+		return
 	}
-}
 
-otelLogWrite "INFO Splunk_TA_otelutils.ps1 started"
-start-sleep -s 3
-if (isOtelProcessRunning($otelProcessName)) {
-	otelLogWrite "INFO Otel agent running"
-	waitForExit($parentPid)
-	forceStopOtelProcess($otelPid)
-	CheckOtelProcessStop($otelPid)
-} else {
-	otelLogWrite "ERROR Otel agent not running"
+	Write-Log "INFO Otel agent running (PID $($otelProcess.ProcessId))"
+	$parentJob      = Start-Job { Wait-Process -Id $using:parentProcess.ProcessId }      -Name "cmd job"
+	$grandParentJob = Start-Job { Wait-Process -Id $using:grandParentProcess.ProcessId } -Name "splunkd job"
+	$otelJob        = Start-Job { Wait-Process -Id $using:otelProcess.ProcessId }        -Name "Otel agent job"
+
+	Write-Log "INFO waiting on termination of splunkd, cmd, or the Otel agent"
+	$finishedJob = Wait-Job -Any -Job $parentJob,$grandParentJob,$otelJob
+
+	Write-Log "INFO at least one of the jobs finished:`n`t'$($finishedJob.Name)': finished with $($finishedJob.JobStateInfo)"
+	Stop-ProcessById $otelProcess.ProcessId
```

Comment:
Out of complete ignorance: Why don't we stop the splunkd job?

Reply:
Stopping a job object in this case would cause the code waiting on the process to end, not the process that we are waiting on - so it doesn't have the desired effect.

Given your Q I notice that, perhaps, it will be good form to call `Remove-Job` on the job objects. However, In this specific case it is not a problem because they will be terminated when powershell exits.